runtime/compiler/env/J9ClassEnv.cpp

```diff
@@ -417,24 +418,79 @@ J9::ClassEnv::isAnonymousClass(TR::Compilation *comp, TR_OpaqueClassBlock *clazz
    return comp->fej9()->isAnonymousClass(clazz);
    }
 
-const TR::TypeLayout*
-J9::ClassEnv::enumerateFields(TR::Region& region, TR_OpaqueClassBlock * opaqueClazz, TR::Compilation *comp)
-   {  
-   J9Class *clazz = (J9Class*)opaqueClazz;
-   TR_VMFieldsInfo fieldsInfo(comp, clazz, 1, stackAlloc);
-   ListIterator<TR_VMField> iter(fieldsInfo.getFields());
-   TR::TypeLayoutBuilder tlb(region);
-   for (TR_VMField *field = iter.getFirst(); field; field = iter.getNext())
+/*
+ * \param prefix
+ *    prefix could be ended with `.` or NULL.
+ * \param prefixLength
+ *    The length of the prefix string.
+ * \param toConcatenateAsPrefix
+ *    If true, the new name is appended with "." to be used as a part of the flattened field chain name.
+ * \param mergedLength
+ *    The merged length of the concatenated field name that is returned to the caller.
+ */
+static char * mergeFieldNames(char * prefix, uint32_t prefixLength, J9ROMFieldShape * field,
+                              bool toConcatenateAsPrefix, TR::Region &region, uint32_t &mergedLength)
+   {
+   J9UTF8 * fieldNameUTF = J9ROMFIELDSHAPE_NAME(field);
+   char * fieldName = reinterpret_cast<char *>(J9UTF8_DATA(fieldNameUTF));
+   uint32_t nameLength = J9UTF8_LENGTH(fieldNameUTF);
+
+   mergedLength = nameLength + prefixLength;
+   if (toConcatenateAsPrefix)
+      mergedLength++;
+   mergedLength++; /* for adding \0 at the end */
+
+   char * newName = new (region) char[mergedLength];
+
+   if (prefixLength > 0)
+      strncpy(newName, prefix, prefixLength);
+   strncpy(newName + prefixLength, fieldName, nameLength);
+
+   if (toConcatenateAsPrefix)
+      newName[mergedLength-2] = '.';
+   newName[mergedLength-1] = '\0';
+
+   return newName;
+   }
+
+static void addEntryForField(TR_VMField * field, TR::TypeLayoutBuilder &tlb, TR::Region& region, J9Class * definingClass,
```

Comment:
can we keep the definingClass as an OpaqueClassBlock and only do the cast internally when passing to the VM APIs to keep consistency in how the JIT code uses the types?

Reply:
the prefix, prefixLength and offsetBase params only seem used with the recursion perhaps create a frontend function to default these with the recrusion taking all the params with an impl suffix to keep usage simple and to make it easy to change defaults in the future...

Reply:
>can we keep the definingClass as an OpaqueClassBlock and only do the cast internally when passing to the VM APIs to keep consistency in how the JIT code uses the types?

`addEntryForField()` is called recursively when the field is flattened. It passes `J9Class *` pointer received from VM to `addEntryForField()` again. If `TR_OpaqueClassBlock *` is used, the pointer needs to be recast back to `TR_OpaqueClassBlock *` again. If we use `J9Class *` as now, it will avoid the recast in the recursive calls. 

``` c++
      J9Class *fieldClass = vm->internalVMFunctions->getFlattenableFieldType(definingClass, field->shape);
      TR_VMFieldsInfo fieldsInfo(comp, fieldClass, 1, stackAlloc);
      ListIterator<TR_VMField> iter(fieldsInfo.getFields());
      for (TR_VMField *childField = iter.getFirst(); childField; childField = iter.getNext())
         {
         addEntryForField(childField, tlb, region, fieldClass, prefixForChild, mergedLength-1, offsetBaseForChild, comp);
         }
```

Reply:
>the prefix, prefixLength and offsetBase params only seem used with the recursion perhaps create a frontend function to default these with the recrusion taking all the params with an impl suffix 

I'm not sure I'm following. Could you elaborate it a bit? Or an example?

Reply:
so apart from the addEntryForField recursive call there is one call elsewhere in the file where the parameters for prefix, prefixLenght and offsetBase seem to be given an initial value - these are updated during the recursion from what I can see. So rather than having the non-recursive call look like addEntryForField(...., 0, 0, NULL) you could create an addEntryForField without those params which supplies those default values to call the current method renamed to addEntryForFieldImpl. That way if we add more dependencies on to it later the defaults are all in one place and not spread around the code base. Does that make sense?

Reply:
For the API issue above splitting the method into addEntryForField and addEntryForFieldImpl would allow the OpaqueClassBlock to J9Class conversion to be done in addEntryForField and for addEntryForFieldImpl to use the J9Class for the beneficial reasons you note.